### PR TITLE
[TRTLLM-14620][feat] Add ragged sparse-attention row contracts

### PR DESCRIPTION
**Pull request by zhaoyuanh-nvidia**
## Description

This is the kernel and sparse-attention foundation for confidence-guided DSpark verification.

It adds token-major ragged row contracts to the DeepSeek-V4 sparse-attention metadata path and extends the compressor and Indexer Top-K operators to consume the corresponding row mappings. Dense/static behavior remains the default when ragged metadata is absent. The current-main port uses the extracted FMHA manager and avoids redundant row-length staging while preserving the post-overlap device refresh.

The change preserves MLA RoPE forwarding and validates fused-Q row shapes, token-major row construction, compressor behavior, and Indexer Top-K behavior. No configuration or test-list files are changed.

## Stack

**Stack position: 1 of 6**

- Umbrella design and performance record: [NVIDIA/TensorRT-LLM#18392](https://github.com/NVIDIA/TensorRT-LLM/pull/18392)
- Base: `main`
- Successor: [zhaoyuanh-nvidia/TensorRT-LLM#1](https://github.com/zhaoyuanh-nvidia/TensorRT-LLM/pull/1)

This is the first independently reviewable layer. Upstream branch creation is protected, so dependent layers are hosted as linked draft fork PRs until their predecessor merges.

## Test Coverage

Focused files include:

- `test_trtllm_mla_rope_forwarding.py`
- `test_deepseek_v4_sparse_mla.py`
- `test_ragged_fused_q_rope_specs.py`
- `test_token_major_rows.py`
- `test_dsa_indexer.py`
- `test_compressor_ragged.py`
- `test_indexer_topk_ragged.py`

- Exact restacked boundary tree: `fa7ac14cdf609659bf1df9a9d79db2d445557dc5`.
- Static boundary checks: `git diff --check` passed; the checked-in formatter and source hooks pass on the cumulative exact tree.
- PR1-focused coverage is included in the exact cumulative 26-selector run: `519 collected`, `515 passed`, `4 skipped`, zero failed.
- Exact current-main cumulative build/runtime: Release SM100+SM103 build, wheel linkage/import/ABI, and changed native-op registration passed on BIA job `817089`.
- Prior-tree evidence only: the pre-restack cumulative PR1 suite reported `158 passed`, `4 skipped`, and `3 subtests passed`; that result does not validate this rewritten tree.
- Cumulative validation is tracked by umbrella PR #18392; the exact-tree result seal is `d8d285aa32a1356feeb311645e4905217887ad9976dd6d956f3ccb852e4552ff`.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work)).
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities.
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) is updated if ownership changes.
- Documentation is updated as needed.
- Update the [TAVA architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in the PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py [hunk 1]
```diff
@@ -205,6 +266,22 @@ def __post_init__(self):
...
 
     def prepare(self):
+        # Guard pinned staging buffers against reuse while the prior step's
+        # non-blocking H2D copies are still queued by the overlap scheduler.
+        if not torch.cuda.is_current_stream_capturing():
+            event = getattr(self, "_prepare_stage_evt", None)
+            if event is None:
+                self._prepare_stage_evt = torch.cuda.Event()
+            else:
+                event.synchronize()
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: New sync point? This will apply to the non-confidence head path too. Is it necessary?

Comment (zhaoyuanh-nvidia) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Good catch. The scope is too broad. I'll replace it with bounded double-buffered staging and per-slot events, so steady-state preparation rotates to a free slot and only waits if both slots are genuinely still in flight or storage must be reallocated.

File: cpp/tensorrt_llm/kernels/compressorKernels/compressorKernels.cu [hunk 2]
```diff
@@ -312,10 +321,16 @@ __global__ void pagedKvCompressKernel(void const* __restrict__ kv_score_raw, flo
...
+    // HEAD_BLOCKS splits head_dim across blockIdx.y for better SM utilisation.
+    //
+    //   HD=512, max element size 2: NTHRD_BASE=64, HEAD_BLOCKS=2,
+    //   NTHRD_INNER=32.
+    //
+    //   HD=128, max element size 2 or 4: NTHRD_BASE=32, HEAD_BLOCKS=1,
+    //   NTHRD_INNER=32.
+    //
+    //   HD=512, max element size 4: NTHRD_BASE=128, HEAD_BLOCKS=4,
+    //   NTHRD_INNER=32.
```
Comment (mikeiovine) [inline cpp/tensorrt_llm/kernels/compressorKernels/compressorKernels.cu]: What's up the comment reformatting in the C++ part of this PR? Can we revert those changes to lower the amount of noise?

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py [hunk 3]
```diff
@@ -39,6 +39,47 @@
 # CuTe DSL Top-K compile key, so warmup must use the runtime dtype.
 _INDEXER_LOGITS_DTYPE = torch.float32
 
+# ``prepare()`` is allowed to run ahead of the GPU under the overlap scheduler.
+# Every tensor below is persistent CPU storage that prepare (or the virtual
+# DeepSeek-V4 implementation it dispatches to) rewrites before a non-blocking
+# H2D copy, or exposes as a host runtime view. Keep the list explicit: blindly
+# cloning every CPU tensor on the metadata object would also duplicate cache
+# manager storage and unrelated, potentially very large, host buffers.
+_PREPARE_HOST_STAGE_FIELDS = (
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: I still don't really get why we need this - we have stuff in model engine that prevents writing to pinned host buffers until the previous H2D copies are finished, I thought.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py [hunk 4]
```diff
@@ -1009,10 +1253,110 @@ def create_expanded_buffers(self, capture_graph=False):
...
+        self.row_req_idx_host = torch.zeros_like(
+            self.row_req_idx_cuda, device="cpu", pin_memory=prefer_pinned()
+        )
+        self._ragged_num_rows = 0
+
+        # Parallel token-major views for MLA RoPE and sparse-MLA generation.
+        # Context requests keep one row; generation contributes one row per
+        # query token. Request-major metadata remains untouched.
+        attn_rows_cap = self.max_num_sequences * (2 + self._draft_sizing_cap)
+        self._attn_num_rows = 0
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Can we gate these new buffers on the confidence head to save on some memory?

**Review by mikeiovine — CHANGES_REQUESTED** [review]

File: cpp/tensorrt_llm/kernels/compressorKernels/compressorKernels.cu (hunk 2 above)
Comment (zhaoyuanh-nvidia) [reply cpp/tensorrt_llm/kernels/compressorKernels/compressorKernels.cu]: These changes were incidental formatting/comment churn. Restored.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py (hunk 3 above)
Comment (zhaoyuanh-nvidia) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Right, I just removed the DSA-local staging ring, CUDA events, synchronization, and redundant H2D copies. The existing `ModelEngine._prepare_inputs_event` and executor ordering already protect the lifetime of the host staging buffers, so adding another metadata-local synchronization mechanism was unnecessary.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py (hunk 4 above)
Comment (zhaoyuanh-nvidia) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa/metadata.py]: Yes, the ragged row-major, token-major, and fused-Q metadata buffers are now guarded by an allocation-time `enable_ragged_verification` capability wired from confidence scheduling. They are not allocated for feature-off static DSA, draft attention, or cross-attention metadata.

File: tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py [hunk 5]
```diff
@@ -125,9 +145,59 @@ class DSAtrtllmAttentionMetadata(TrtllmAttentionMetadata):
...
+    ragged_verify_lens: Optional[List[int]] = None
+    # Allocate the row-major and token-major buffers used only by ragged
+    # verification. This is fixed before CUDA-graph metadata is created.
+    enable_ragged_verification: bool = False
+    # In device-window mode the host split determines only the captured shape;
+    # the true windows are installed before replay.
+    device_windows_mode: bool = False
+    # Query tokens per generation request for a uniform runtime tier. Zero
+    # falls back to the static draft-token ceiling.
+    runtime_tokens_per_gen_step: int = 0
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py]: Where is `enable_ragged_verification` set to true? Seems like a lot of the code introduced in this PR is not reachable

File: tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/metadata.py [hunk 6]
```diff
@@ -549,6 +619,17 @@ def prepare(self):
...
+        if self.device_windows_mode and self.is_ragged_verify:
+            # The host split only chooses the captured shape.  Until the
+            # device prologue installs the freshly ranked windows, every
+            # host-side consumer must see a full-window upper bound rather
+            # than stale per-request shape values.  Keep this identical to the
+            # native DSA metadata contract.
+            nc, ns = self.num_contexts, self.num_seqs
+            kv_lens = kv_lens.clone()
+            kv_lens[nc:ns] += (1 + self.max_draft_tokens) - self.seq_lens_kv[nc:ns]
+            self.kv_lens[nc:ns] = kv_lens[nc:ns]
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention/backends/sparse/deepseek_v4/metadata.py:631]: Could probably put this into a helper function, identical code in `dsa/metadata.py`

File: tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py [hunk 7]
```diff
@@ -125,9 +145,59 @@ class DSAtrtllmAttentionMetadata(TrtllmAttentionMetadata):
...
+
+    @property
+    def is_ragged_verify(self) -> bool:
+        if self.ragged_verify_lens is not None and not self.enable_ragged_verification:
+            raise RuntimeError(
+                "ragged verification metadata was installed without enabling "
+                "ragged-verification buffers"
+            )
+        return self.ragged_verify_lens is not None
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py:168]: nit; avoid properties that raise exceptions. Can we do this verification when this object is initialized?

File: tests/unittest/_torch/attention/sparse/deepseek_v4/test_ragged_fused_q_rope_specs.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+            num_generations=2,
+            seq_lens_cuda=_Tensor([1, 2]),
+            fused_q_gen_cu_seqlens=None,
+            _fused_q_gen_cu_seqlens_valid=False,
+        )
+        self.assertIsNone(_PREP_GEN_PREFIX(metadata))
+        metadata.fused_q_gen_cu_seqlens = _Tensor([0, 0])
+        self.assertIsNone(_PREP_GEN_PREFIX(metadata))
+
+    def test_storage_and_rebuild_are_cuda_graph_structural(self):
```
Comment (mikeiovine) [inline tests/unittest/_torch/attention/sparse/deepseek_v4/test_ragged_fused_q_rope_specs.py:321]: Tests like this (i.e. those that use `ast` to parse a module and verify the presence/absence of various methods and classes seem very fragile. What value are we adding? I'd consider removing these if the value is low

File: cpp/tensorrt_llm/kernels/compressorKernels/compressorKernels.h [hunk 9]
```diff
@@ -47,6 +47,11 @@ void pagedKvCompressLaunch(void const* kv_score, // [m, 2*state_dim]  (bf16 or f
     int kv_score_elem_bytes,                     // bytes per element for kv_score (2=bf16, 4=fp32)
     int state_elem_bytes,                        // bytes per element for paged state (2=bf16, 4=fp32)
     int out_elem_bytes,                          // bytes per element for output
+    int32_t const* new_tokens_per_seq,           // [bsz] or nullptr; per-request tokens appended
+                                                 // this step. nullptr means every request appended
+                                                 // next_n (the uniform case); non-null makes next_n
+                                                 // an upper bound only, for DSpark ragged
```
Comment (mikeiovine) [inline cpp/tensorrt_llm/kernels/compressorKernels/compressorKernels.h]: Can probably remove "DSpark" in this comment

File: cpp/tensorrt_llm/kernels/IndexerTopK.h [hunk 10]
```diff
@@ -38,10 +38,20 @@ int computeIndexerTopKDecodeBlocksPerRow(int numRows, int numColumns, int splitW
...
+///
+/// @param rowKvLens  optional [numRows] per-row causal KV extent, in
+/// uncompressed
+///   token space. Non-null selects the ragged layout: rows no longer map to
+///   requests by `rowIdx / next_n`, so the extent every row may attend to is
+///   supplied directly instead of being reconstructed from `next_n`. Null keeps
+///   the uniform arithmetic, bit-identical to before. A
+///   negative extent or one whose compressed length exceeds `numColumns` fails
+///   closed as an empty row, so device-updated graph inputs cannot read beyond
+///   the corresponding logits row.
```
Comment (mikeiovine) [inline cpp/tensorrt_llm/kernels/IndexerTopK.h]: Formatting looks a bit weird here

File: cpp/tensorrt_llm/kernels/IndexerTopK.h [hunk 11]
```diff
@@ -38,10 +38,20 @@ int computeIndexerTopKDecodeBlocksPerRow(int numRows, int numColumns, int splitW
 ///   - Insertion sort   (N < kSortingAlgorithmThreshold)
 ///   - Radix sort       (kSortingAlgorithmThreshold ≤ N < splitWork)
 ///   - Radix split-work (N ≥ splitWork — uses outLogitsAux / outIndicesAux)
+///
+/// @param rowKvLens  optional [numRows] per-row causal KV extent, in
+/// uncompressed
+///   token space. Non-null selects the ragged layout: rows no longer map to
+///   requests by `rowIdx / next_n`, so the extent every row may attend to is
+///   supplied directly instead of being reconstructed from `next_n`. Null keeps
+///   the uniform arithmetic, bit-identical to before. A
```
Comment (mikeiovine) [inline cpp/tensorrt_llm/kernels/IndexerTopK.h]: "bit-identical to before" won't make much sense outside the context of this PR, can probably remove that part of the sentence

File: tensorrt_llm/_torch/attention/mla.py [hunk 12]
```diff
@@ -1531,8 +1531,21 @@ def forward_absorption_generation(
...
+        # The rope kernel writes seqQOffset[batch_idx + 1] for every
+        # batch_idx < batch_size, and CUB-scans total_s_len / seq_len entries
+        # into seqKVOffsets (mlaKernels.cu:579, :650). `batch_size` there is the
+        # count of presented rows, which equals num_seqs only while every
+        # request contributes the same number of query tokens. Under the
+        # token-major presentation a row is a token, so size these by the token
+        # count. Uniform verification keeps the established request-major
+        # allocation rather than scaling these temporary buffers by K.
+        cu_seqlens_rows = (
+            max(num_seqs, num_tokens) + 1
```
Comment (mikeiovine) [inline tensorrt_llm/_torch/attention/mla.py]: Do we need the `max()` here? Not sure I follow the comment. How can `num_tokens` be less than `num_seqs`?

**Review by mikeiovine — APPROVED** [review]
Only nitpicks now. Looks good
